tests/test_kb_auth.py

```diff
@@ -0,0 +1,123 @@
+import asyncio
+import pytest
+from staging_service.kb_auth_client import InvalidUserError, KBaseAuth, InvalidTokenError, _get
+from tests.test_utils import bootstrap_config
+import os
+from unittest.mock import patch
+
+config = bootstrap_config()
+AUTH_URL = config["staging_service"]["AUTH_URL"]
+
+# NOTE: These tests are intended to be run against the auth service at
+# https://ci.kbase.us/services/auth
+# Any unexpected failures should check the auth service, the given test
+# token, and whether the test user exists or not.
+
+# TODO: add env var names to test config
+TEST_TOKEN = os.environ.get("KBASE_TEST_TOKEN")
+TEST_USER = os.environ.get("KBASE_TEST_USER")
+VALID_TEST_USER = "narrativetest"
+NOT_REAL_USER = "please_do_not_ever_make_this_username_what_is_wrong_with_you"
+INVALID_USER = "_(__)_"
+
+
+@pytest.fixture
+async def auth_client():
+    """
+    A fixture that makes a default auth client with a shiny new cache.
+    Avoids the extra lag from using KBaseAuth.create.
+    """
+    return await KBaseAuth.create(AUTH_URL)
+
+
+async def test_create():
+    auth_client = await KBaseAuth.create(AUTH_URL)
+    assert isinstance(auth_client, KBaseAuth)
+
+
+async def test_create_non_auth_url():
+    with pytest.raises(IOError, match="does not appear to be the KBase Authentication Service"):
+        await KBaseAuth.create("https://github.com")
+
+
+async def test_create_incorrect_url():
+    with pytest.raises(IOError, match="Error from KBase auth server"):
+        await KBaseAuth.create(AUTH_URL + "wrong")
+
+
+async def test_non_json_error():
+    with pytest.raises(IOError, match="Non-JSON response from KBase auth server"):
+        await KBaseAuth.create("https://kbase.us/services")
+
+
+# TODO: test other service fail modes
+
+
+async def test_get_user(auth_client: KBaseAuth):
+    assert await auth_client.get_user(TEST_TOKEN) == TEST_USER
+
+
+async def test_get_user_from_cache(auth_client: KBaseAuth):
+    # patch this to act as a spy, should only be called once with the caching done.
+    with patch("staging_service.kb_auth_client._get", wraps=_get) as spy_get:
+        result = await auth_client.get_user(TEST_TOKEN)
+        assert result == TEST_USER
+
+        result = await auth_client.get_user(TEST_TOKEN)
+        assert result == TEST_USER
+
```

Comment:
For some reason codecov is saying this isn't exercising the cache

Reply:
It's also saying neither of the 2 auth client API methods ever return, which is odd

Reply:
I'm also not seeing codecov's complaints anymore in the check results. It has different complaints that'll get addressed later.

Reply:
Hmm, the annotations are gone from the auth client changes UI, but the codecov site still says there are uncovered lines

Reply:
Codecov hasn't edited its PR comment in 4 days